tests/generators/generate_hfdata.py

```diff
@@ -0,0 +1,125 @@
+from tests.generators.dump_pyscf import dump_pyscf
+from tests import testcases
+
+from pyscf import gto, scf
+
+from pathlib import Path
+import h5py
+import os
+import tempfile
+
+
+_testdata_dirname = "data"
+
+
+def run_pyscf(test_case: testcases.TestCase, frac_occ: bool):
+    """
+    Runs the pyscf calculation for the given testcase.
+    """
+    # create a temporary directory for the temporary pyscf files:
+    # on a cluster the /tmp folder might not be cleaned as often.
+    with tempfile.TemporaryDirectory(dir=Path.cwd()) as tmpdir:
+        old_tmpdir = os.environ.get("PYSCF_TMPDIR", None)
+        os.environ["PYSCF_TMPDIR"] = tmpdir
+        # Run SCF in pyscf and converge super-tight using an EDIIS
+        mol = gto.M(
+            atom=test_case.xyz,
+            basis=test_case.basis,
+            unit=test_case.unit,
+            spin=test_case.multiplicity - 1,  # =2S
+            verbose=0
+        )
+        mf = scf.RHF(mol) if test_case.restricted else scf.UHF(mol)
+        mf.diis = scf.EDIIS()
+        mf.conv_tol = 1e-14
+        mf.conv_tol_grad = 1e-12
+        mf.diis_space = 6
+        mf.max_cycle = 600
+        if frac_occ:
+            mf = scf.addons.frac_occ(mf)
+        mf.kernel()
+        # restore the original state of the env variable
+        if old_tmpdir is None:
+            del os.environ["PYSCF_TMPDIR"]
+        else:
+            os.environ["PYSCF_TMPDIR"] = old_tmpdir
+        assert mf.converged  # ensure that the SCF is converged
+    return mf
+
+
+def generate(test_case: testcases.TestCase, frac_occ: bool) -> h5py.File:
+    """
+    Run Pyscf for the given test case and dump the result in the hdf5 file
+    if the file does not already exist.
+    """
+    hdf5_file = Path(__file__).resolve().parent.parent / _testdata_dirname
+    hdf5_file /= test_case.hfdata_file_name
+    if hdf5_file.exists():
+        return None
+    print(f"Generating data for {test_case.file_name}.")
+    mf = run_pyscf(test_case, frac_occ)
+    hdf5_file = h5py.File(hdf5_file, "w")
+    dump_pyscf(mf, hdf5_file)
+    return hdf5_file
+
+
+def generate_ch2nh2():
+    cases = testcases.get(n_expected_cases=1, name="ch2nh2", basis="sto-3g").pop()
```

Comment:
Why is `n_expected_cases` needed?

Reply:
The idea is to raise an exception if another `TestCase` that matches your request is added. For instance you want to retrieve the `h2o_sto3g` and `h2o_def2tzvp` cases with `testcases.get(name="h2o")`, but later `h2o_ccpvdz` is added and now the request will return all 3 cases, which is something you might not want at this point.
The `get` function is from an earlier iteration anyway. Most of the code is now using `get_by_filename`, which similarly raises an exception if it can't find a `TestCase` for each filename.